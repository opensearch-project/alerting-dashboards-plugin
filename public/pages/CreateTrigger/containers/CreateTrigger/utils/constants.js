/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const TRIGGER_TYPE = {
  AD: 'anomaly_detector_trigger',
  BUCKET_LEVEL: 'bucket_level_trigger',
  ALERT_TRIGGER: 'alerting_trigger',
  QUERY_LEVEL: 'query_level_trigger',
  DOC_LEVEL: 'document_level_trigger',
};

export const FORMIK_INITIAL_BUCKET_SELECTOR_VALUES = {
  buckets_path: {
    '': '',
  },
  parent_bucket_path: 'composite_agg',
  script: {
    source: 'params.',
  },
};

export const FORMIK_INITIAL_TRIGGER_CONDITION_VALUES = {
  thresholdValue: 10000,
  thresholdEnum: 'ABOVE',
  anomalyDetector: {
    triggerType: TRIGGER_TYPE.AD,
    anomalyGradeThresholdValue: 0.7,
    anomalyGradeThresholdEnum: 'ABOVE',
    anomalyConfidenceThresholdValue: 0.7,
    anomalyConfidenceThresholdEnum: 'ABOVE',
  },
  script: {
    lang: 'painless',
    source: 'ctx.results[0].hits.total.value > 0',
  },
  buckets_path: {},
  parent_bucket_path: 'composite_agg',
  gap_policy: '',
  query: undefined,
  queryMetric: undefined,
  andOrCondition: undefined,
};

export const FORMIK_INITIAL_TRIGGER_VALUES = {
  name: '',
  severity: '1',
  minTimeBetweenExecutions: null,
  rollingWindowSize: null,
  script: {
    lang: 'painless',
    source: `ctx.results[0].hits.total.value > 0`,
  },
  bucketSelector: JSON.stringify(FORMIK_INITIAL_BUCKET_SELECTOR_VALUES, null, 4), // TODO: To be used for Bucket-Level Triggers defined by query
  triggerConditions: [],
  thresholdValue: 10000,
  thresholdEnum: 'ABOVE',
  anomalyDetector: {
    triggerType: TRIGGER_TYPE.AD,
    anomalyGradeThresholdValue: 0.7,
    anomalyGradeThresholdEnum: 'ABOVE',
    anomalyConfidenceThresholdValue: 0.7,
    anomalyConfidenceThresholdEnum: 'ABOVE',
  },
  where: {
    fieldName: [],
    operator: 'includes',
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  actions: undefined,
};

export const FORMIK_INITIAL_DOC_LEVEL_SCRIPT = {
  lang: FORMIK_INITIAL_TRIGGER_VALUES.script.lang,
  source: '(query[name=<queryName>] || query[name=<queryName>]) && query[tag=<queryTag>]',
};

export const HITS_TOTAL_RESULTS_PATH = 'ctx.results[0].hits.total.value';
export const AGGREGATION_RESULTS_PATH = 'ctx.results[0].aggregations.metric.value';
export const ANOMALY_GRADE_RESULT_PATH = 'ctx.results[0].aggregations.max_anomaly_grade.value';
export const ANOMALY_CONFIDENCE_RESULT_PATH = 'ctx.results[0].hits.hits[0]._source.confidence';
export const NOT_EMPTY_RESULT =
  'ctx.results != null && ctx.results.length > 0 && ctx.results[0].aggregations != null && ctx.results[0].aggregations.max_anomaly_grade != null && ctx.results[0].hits.total.value > 0 && ctx.results[0].hits.hits[0]._source != null && ctx.results[0].hits.hits[0]._source.confidence != null';
