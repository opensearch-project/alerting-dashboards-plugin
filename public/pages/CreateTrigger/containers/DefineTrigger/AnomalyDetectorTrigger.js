/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import _ from 'lodash';
import { EuiSpacer } from '@elastic/eui';
import { AnomalyDetectorData } from '../../../CreateMonitor/containers/AnomalyDetectors/AnomalyDetectorData';
import TriggerExpressions from '../../components/TriggerExpressions';
import { AnomaliesChart } from '../../../CreateMonitor/components/AnomalyDetectors/AnomaliesChart';
import { EmptyFeaturesMessage } from '../../../CreateMonitor/components/AnomalyDetectors/EmptyFeaturesMessage/EmptyFeaturesMessage';
import { EmptyDetectorMessage } from '../../../CreateMonitor/components/AnomalyDetectors/EmptyDetectorMessage/EmptyDetectorMessage';

class AnomalyDetectorTrigger extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    const { adValues, detectorId, fieldPath, flyoutMode } = this.props;
    return (
      <div style={flyoutMode ? {} : { padding: '0px 10px' }}>
        <AnomalyDetectorData
          detectorId={detectorId}
          render={(anomalyData) => {
            // using lodash.get without worrying about whether an intermediate property is null or undefined.
            if (_.get(anomalyData, 'anomalyResult.anomalies', []).length > 0) {
              return (
                <React.Fragment>
                  <TriggerExpressions
                    thresholdValue={adValues.anomalyGradeThresholdValue}
                    thresholdEnum={adValues.anomalyGradeThresholdEnum}
                    keyFieldName={`${fieldPath}anomalyDetector.anomalyGradeThresholdEnum`}
                    valueFieldName={`${fieldPath}anomalyDetector.anomalyGradeThresholdValue`}
                    label="Anomaly grade threshold"
                  />
                  {!flyoutMode && (
                    <>
                      <EuiSpacer size="xs" />
                      <AnomaliesChart
                        showTitle={false}
                        showSettings={false}
                        startDateTime={anomalyData.previewStartTime}
                        endDateTime={anomalyData.previewEndTime}
                        anomalies={anomalyData.anomalyResult.anomalies}
                        isLoading={anomalyData.isLoading}
                        displayGrade
                        annotationData={[{ dataValue: adValues.anomalyGradeThresholdValue }]}
                      />
                    </>
                  )}
                  <EuiSpacer size="m" />
                  <TriggerExpressions
                    thresholdValue={adValues.anomalyConfidenceThresholdValue}
                    thresholdEnum={adValues.anomalyConfidenceThresholdEnum}
                    keyFieldName={`${fieldPath}anomalyDetector.anomalyConfidenceThresholdEnum`}
                    valueFieldName={`${fieldPath}anomalyDetector.anomalyConfidenceThresholdValue`}
                    label="Anomaly confidence threshold"
                  />
                  {!flyoutMode && (
                    <>
                      <EuiSpacer size="xs" />
                      <AnomaliesChart
                        showTitle={false}
                        showSettings={false}
                        startDateTime={anomalyData.previewStartTime}
                        endDateTime={anomalyData.previewEndTime}
                        anomalies={anomalyData.anomalyResult.anomalies}
                        isLoading={anomalyData.isLoading}
                        displayConfidence
                        annotationData={[{ dataValue: adValues.anomalyConfidenceThresholdValue }]}
                      />
                    </>
                  )}
                </React.Fragment>
              );
            } else {
              return _.isEmpty(detectorId) ? (
                <EmptyDetectorMessage />
              ) : (
                <EmptyFeaturesMessage detectorId={detectorId} isLoading={anomalyData.isLoading} />
              );
            }
          }}
        />
      </div>
    );
  }
}

export { AnomalyDetectorTrigger };
