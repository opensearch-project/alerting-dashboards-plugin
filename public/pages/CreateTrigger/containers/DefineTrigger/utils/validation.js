/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const validateTriggerName = (triggers = [], index, isFlyOut = false) => {
  return (value) => {
    const trimmedValue = value.trim();
    if (!trimmedValue) return isFlyOut ? 'Required.' : 'Trigger name is required.';
    const triggerNameExistWithIndex = triggers.some((trigger, i) => {
      return i !== index && trimmedValue === trigger.name.trim();
    });
    if (triggerNameExistWithIndex) {
      return 'Trigger name already used.';
    }

    // TODO: character restrictions
    // TODO: character limits
  };
};
