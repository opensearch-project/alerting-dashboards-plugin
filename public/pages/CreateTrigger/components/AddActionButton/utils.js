/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { DEFAULT_MESSAGE_SOURCE, FORMIK_INITIAL_ACTION_VALUES } from '../../utils/constants';
import { MONITOR_TYPE } from '../../../../utils/constants';
import { getDigitId, getUniqueName } from '../../../../utils/helpers';

export const getInitialActionValues = ({ monitorType, flyoutMode, actions }) => {
  const initialActionValues = _.cloneDeep(FORMIK_INITIAL_ACTION_VALUES);

  switch (monitorType) {
    case MONITOR_TYPE.BUCKET_LEVEL:
      _.set(
        initialActionValues,
        'message_template.source',
        DEFAULT_MESSAGE_SOURCE.BUCKET_LEVEL_MONITOR
      );
      break;
    case MONITOR_TYPE.QUERY_LEVEL:
    case MONITOR_TYPE.CLUSTER_METRICS:
      _.set(
        initialActionValues,
        'message_template.source',
        DEFAULT_MESSAGE_SOURCE.QUERY_LEVEL_MONITOR
      );
      break;
  }

  const id = getDigitId();
  initialActionValues.id = `notification${id}`;

  // Add name based on previous name;
  if (flyoutMode) {
    initialActionValues.name = getUniqueName(actions, 'Notification ');
  }

  return initialActionValues;
};
