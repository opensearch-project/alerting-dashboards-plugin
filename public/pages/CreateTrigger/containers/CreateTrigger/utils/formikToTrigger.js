/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import {
  AGGREGATION_RESULTS_PATH,
  ANOMALY_CONFIDENCE_RESULT_PATH,
  ANOMALY_GRADE_RESULT_PATH,
  FORMIK_INITIAL_TRIGGER_VALUES,
  HITS_TOTAL_RESULTS_PATH,
  NOT_EMPTY_RESULT,
  TRIGGER_TYPE,
} from './constants';
import { MONITOR_TYPE, SEARCH_TYPE } from '../../../../../utils/constants';
import { NOTIFY_OPTIONS_VALUES } from '../../../components/Action/actions/Message';
import { FORMIK_INITIAL_ACTION_VALUES } from '../../../utils/constants';

export function formikToTrigger(values, monitorUiMetadata = {}) {
  const triggerDefinitions = _.get(values, 'triggerDefinitions');
  return _.isArray(triggerDefinitions)
    ? formikToTriggerDefinitions(triggerDefinitions, monitorUiMetadata)
    : formikToTriggerDefinition(values, monitorUiMetadata);
}

export function formikToTriggerDefinitions(values, monitorUiMetadata) {
  return values.map((trigger) => formikToTriggerDefinition(trigger, monitorUiMetadata));
}

export function formikToTriggerDefinition(values, monitorUiMetadata) {
  switch (monitorUiMetadata.monitor_type) {
    case MONITOR_TYPE.BUCKET_LEVEL:
      return formikToBucketLevelTrigger(values, monitorUiMetadata);
    case MONITOR_TYPE.DOC_LEVEL:
      return formikToDocumentLevelTrigger(values, monitorUiMetadata);
    default:
      return formikToQueryLevelTrigger(values, monitorUiMetadata);
  }
}

export function formikToQueryLevelTrigger(values, monitorUiMetadata) {
  const condition = formikToCondition(values, monitorUiMetadata);
  const actions = formikToAction(values);
  // TODO: We probably also want to wrap this with 'query_level_trigger' after
  //  confirming what will break in the frontend (when accessing the fields)
  return {
    id: values.id,
    name: values.name,
    severity: values.severity,
    condition,
    actions: actions,
    min_time_between_executions: values.minTimeBetweenExecutions,
    rolling_window_size: values.rollingWindowSize,
  };
}

export function formikToBucketLevelTrigger(values, monitorUiMetadata) {
  const condition = formikToBucketLevelTriggerCondition(values, monitorUiMetadata);
  const actions = formikToBucketLevelTriggerAction(values);
  return {
    bucket_level_trigger: {
      id: values.id,
      name: values.name,
      severity: values.severity,
      condition,
      actions: actions,
      min_time_between_executions: values.minTimeBetweenExecutions,
      rolling_window_size: values.rollingWindowSize,
    },
  };
}

export function formikToDocumentLevelTrigger(values, monitorUiMetadata) {
  const condition = formikToDocumentLevelTriggerCondition(values, monitorUiMetadata);
  const actions = formikToBucketLevelTriggerAction(values);
  return {
    document_level_trigger: {
      id: values.id,
      name: values.name,
      severity: values.severity,
      condition: condition,
      actions: actions,
    },
  };
}

export function formikToDocumentLevelTriggerCondition(values, monitorUiMetadata) {
  const triggerConditions = _.get(values, 'triggerConditions', []);
  const searchType = _.get(monitorUiMetadata, 'search.searchType', SEARCH_TYPE.QUERY);
  if (searchType === SEARCH_TYPE.QUERY) return { script: values.script };
  const source = getDocumentLevelScriptSource(triggerConditions);
  return {
    script: {
      lang: 'painless',
      source: source,
    },
  };
}

export function getDocumentLevelScriptSource(conditions) {
  const scriptSourceContents = [];
  conditions.forEach((condition) => {
    const { andOrCondition, query } = condition;
    if (andOrCondition) {
      const logicOperator = getLogicalOperator(andOrCondition);
      scriptSourceContents.push(logicOperator);
    }
    if (!_.isEmpty(query) && !_.isEmpty(query.queryName)) {
      const queryExpression = _.get(query, 'expression');
      scriptSourceContents.push(`query[${queryExpression}]`);
    }
  });
  return scriptSourceContents.join(' ');
}

export function formikToAction(values) {
  const actions = values.actions;
  if (actions && actions.length > 0) {
    return actions.map((action) => {
      if (!action.throttle_enabled) return _.omit(action, ['throttle']);
      return action;
    });
  }
  return actions;
}

export function formikToBucketLevelTriggerAction(values) {
  const actions = values.actions;
  const executionPolicyPath = 'action_execution_policy.action_execution_scope';
  if (actions && actions.length > 0) {
    return actions.map((action) => {
      let formattedAction = _.cloneDeep(action);

      switch (formattedAction.throttle_enabled) {
        case true:
          _.set(formattedAction, 'throttle.unit', FORMIK_INITIAL_ACTION_VALUES.throttle.unit);
          break;
        case false:
          formattedAction = _.omit(formattedAction, ['throttle']);
          break;
      }

      const notifyOption = _.get(formattedAction, `${executionPolicyPath}`);
      const notifyOptionId = _.isString(notifyOption) ? notifyOption : _.keys(notifyOption)[0];
      switch (notifyOptionId) {
        case NOTIFY_OPTIONS_VALUES.PER_ALERT:
          const actionableAlerts = _.get(
            formattedAction,
            `${executionPolicyPath}.${NOTIFY_OPTIONS_VALUES.PER_ALERT}.actionable_alerts`,
            []
          );
          _.set(
            formattedAction,
            `${executionPolicyPath}.${NOTIFY_OPTIONS_VALUES.PER_ALERT}.actionable_alerts`,
            actionableAlerts.map((entry) => entry.value)
          );
          break;
        case NOTIFY_OPTIONS_VALUES.PER_EXECUTION:
          _.set(
            formattedAction,
            `${executionPolicyPath}.${NOTIFY_OPTIONS_VALUES.PER_EXECUTION}`,
            {}
          );
          break;
      }
      return formattedAction;
    });
  }
  return actions;
}

export function formikToTriggerUiMetadata(values, monitorUiMetadata) {
  switch (monitorUiMetadata.monitor_type) {
    case MONITOR_TYPE.QUERY_LEVEL:
    case MONITOR_TYPE.CLUSTER_METRICS:
      const searchType = _.get(monitorUiMetadata, 'search.searchType', 'query');
      const queryLevelTriggersUiMetadata = {};
      _.get(values, 'triggerDefinitions', []).forEach((trigger) => {
        const { anomalyDetector, thresholdEnum, thresholdValue } = trigger;
        const triggerMetadata = { value: thresholdValue, enum: thresholdEnum };

        //Store AD values only if AD trigger.
        if (searchType === SEARCH_TYPE.AD && anomalyDetector.triggerType === TRIGGER_TYPE.AD) {
          triggerMetadata.adTriggerMetadata = {
            triggerType: anomalyDetector.triggerType,
            anomalyGrade: {
              value: anomalyDetector.anomalyGradeThresholdValue,
              enum: anomalyDetector.anomalyGradeThresholdEnum,
            },
            anomalyConfidence: {
              value: anomalyDetector.anomalyConfidenceThresholdValue,
              enum: anomalyDetector.anomalyConfidenceThresholdEnum,
            },
          };
        }

        queryLevelTriggersUiMetadata[trigger.name] = triggerMetadata;
      });
      return queryLevelTriggersUiMetadata;

    case MONITOR_TYPE.BUCKET_LEVEL:
      const bucketLevelTriggersUiMetadata = {};
      _.get(values, 'triggerDefinitions', []).forEach((trigger) => {
        const triggerMetadata = trigger.triggerConditions.map((condition) => ({
          value: condition.thresholdValue,
          enum: condition.thresholdEnum,
        }));
        bucketLevelTriggersUiMetadata[trigger.name] = triggerMetadata;
      });
      return bucketLevelTriggersUiMetadata;
    case MONITOR_TYPE.DOC_LEVEL:
      const docLevelTriggersUiMetadata = {};
      _.get(values, 'triggerDefinitions', []).forEach((trigger) => {
        const triggerMetadata = _.get(trigger, 'triggerConditions', []).map((condition) => ({
          query: condition.query,
          andOrCondition: condition.andOrCondition,
          script: condition.script,
        }));
        docLevelTriggersUiMetadata[trigger.name] = triggerMetadata;
      });
      return docLevelTriggersUiMetadata;
  }
}

export function formikToCondition(values, monitorUiMetadata = {}) {
  const { thresholdValue, thresholdEnum } = values;
  const searchType = _.get(monitorUiMetadata, 'search.searchType', 'query');
  const aggregationType = _.get(monitorUiMetadata, 'search.aggregations.0.aggregationType');

  if (searchType === SEARCH_TYPE.QUERY || searchType === SEARCH_TYPE.CLUSTER_METRICS)
    return { script: values.script };
  if (searchType === SEARCH_TYPE.AD) return getADCondition(values);

  // If no aggregation type defined, default to count of documents situation
  const isCount = !aggregationType;
  const resultsPath = getResultsPath(isCount);
  const operator = getRelationalOperator(thresholdEnum);
  return getCondition(resultsPath, operator, thresholdValue, isCount);
}

export function formikToBucketLevelTriggerCondition(values, monitorUiMetadata = {}) {
  const searchType = _.get(monitorUiMetadata, 'search.searchType', SEARCH_TYPE.QUERY);

  let bucketSelector = _.get(
    values,
    'bucketSelector',
    FORMIK_INITIAL_TRIGGER_VALUES.bucketSelector
  );
  try {
    // JSON.parse() throws an exception when the argument is a malformed JSON string.
    // This caused exceptions when tinkering with the JSON in the code editor.
    // This try/catch block will only parse the JSON string if it is not malformed.
    // It will otherwise store the JSON as a string for continued editing.
    bucketSelector = JSON.parse(bucketSelector);
  } catch (err) {}

  if (searchType === SEARCH_TYPE.QUERY) return bucketSelector;
  if (searchType === SEARCH_TYPE.GRAPH) return getBucketLevelTriggerCondition(values);
}

export function getADCondition(values) {
  const { anomalyDetector } = values;
  if (anomalyDetector.triggerType === TRIGGER_TYPE.AD) {
    const anomalyGradeOperator = getRelationalOperator(anomalyDetector.anomalyGradeThresholdEnum);
    const anomalyConfidenceOperator = getRelationalOperator(
      anomalyDetector.anomalyConfidenceThresholdEnum
    );
    return {
      script: {
        lang: 'painless',
        source: `return ${NOT_EMPTY_RESULT} && ${ANOMALY_GRADE_RESULT_PATH} != null && ${ANOMALY_GRADE_RESULT_PATH} ${anomalyGradeOperator} ${anomalyDetector.anomalyGradeThresholdValue} && ${ANOMALY_CONFIDENCE_RESULT_PATH} ${anomalyConfidenceOperator} ${anomalyDetector.anomalyConfidenceThresholdValue}`,
      },
    };
  } else {
    return { script: values.script };
  }
}

export function getCondition(resultsPath, operator, value, isCount) {
  const baseSource = `${resultsPath} ${operator} ${value}`;
  return {
    script: {
      lang: 'painless',
      source: isCount ? baseSource : `return ${resultsPath} == null ? false : ${baseSource}`,
    },
  };
}

export function getBucketLevelTriggerCondition(values) {
  const conditions = values.triggerConditions;
  const bucketsPath = getBucketSelectorBucketsPath(conditions);
  const scriptSource = getBucketSelectorScriptSource(conditions);
  const composite_agg_filter = getCompositeAggFilter(values);

  return {
    parent_bucket_path: 'composite_agg',
    buckets_path: bucketsPath,
    script: {
      source: scriptSource,
    },
    composite_agg_filter: composite_agg_filter,
  };
}

export function getBucketSelectorBucketsPath(conditions) {
  const bucketsPath = {};
  conditions.forEach((condition) => {
    const { queryMetric } = condition;
    bucketsPath[queryMetric] = queryMetric;
  });
  return bucketsPath;
}

export function getBucketSelectorScriptSource(conditions) {
  const scriptSourceContents = [];
  conditions.forEach((condition) => {
    const { queryMetric, thresholdValue, thresholdEnum, andOrCondition } = condition;
    if (andOrCondition) {
      // TODO: If possible, adding parentheses around the AND statements of the resulting script
      //  would improve readability but it shouldn't affect the logical result
      const logicalOperator = getLogicalOperator(andOrCondition);
      scriptSourceContents.push(logicalOperator);
    }
    const relationalOperator = getRelationalOperator(thresholdEnum);
    const scriptCondition = `params.${queryMetric} ${relationalOperator} ${thresholdValue}`;
    scriptSourceContents.push(scriptCondition);
  });
  return scriptSourceContents.join(' ');
}

export function getResultsPath(isCount) {
  return isCount ? HITS_TOTAL_RESULTS_PATH : AGGREGATION_RESULTS_PATH;
}

export function getCompositeAggFilter({ where }) {
  const fieldName = _.get(where, 'fieldName', FORMIK_INITIAL_TRIGGER_VALUES.where.fieldName);
  const composite_agg_filter = {};
  if (fieldName.length > 0) {
    composite_agg_filter[where.fieldName[0].label] = {
      [where.operator]: where.fieldValue,
    };
    return composite_agg_filter;
  }
}

export function getRelationalOperator(thresholdEnum) {
  return { ABOVE: '>', BELOW: '<', EXACTLY: '==' }[thresholdEnum];
}

export function getLogicalOperator(logicalEnum) {
  return { AND: '&&', OR: '||' }[logicalEnum];
}
