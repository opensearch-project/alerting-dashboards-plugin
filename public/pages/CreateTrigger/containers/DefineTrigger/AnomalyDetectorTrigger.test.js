/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { Formik } from 'formik';
import {
  notificationServiceMock,
  httpServiceMock,
} from '../../../../../../../src/core/public/mocks';
import { AnomalyDetectorTrigger } from './AnomalyDetectorTrigger';
import { httpClientMock } from '../../../../../test/mocks';
import { CoreContext } from '../../../../../public/utils/CoreContext';
import { setClient, setNotifications } from '../../../../services';

// enabling waiting until all of the promiseds have cleared: https://tinyurl.com/5hym6n9b
const runAllPromises = () => new Promise(setImmediate);

beforeEach(() => {
  jest.clearAllMocks();
});

describe('AnomalyDetectorTrigger', () => {
  const notifications = notificationServiceMock.createStartContract();
  setNotifications(notifications);
  const httpClientMock = httpServiceMock.createStartContract();
  setClient(httpClientMock);

  test('renders no feature', () => {
    const component = <AnomalyDetectorTrigger detectorId="tempId" />;
    expect(render(component)).toMatchSnapshot();
  });
  test('renders no detector id', () => {
    const component = <AnomalyDetectorTrigger />;
    expect(render(component)).toMatchSnapshot();
  });
  test('renders preview sparse data', async () => {
    // using it since it will render React.Fragment that rendering AnomalyDetectorTrigger returns
    const response = {
      anomalyResult: {
        anomalies: [],
        featureData: {},
      },
      detector: {
        featureAttributes: [
          {
            featureId: 'TV6fFYYB7j86MXY_Bzh2',
            featureName: 'time',
            featureEnabled: true,
            aggregationQuery: {
              time: {
                max: {
                  field: 'time',
                },
              },
            },
          },
        ],
      },
    };

    // Mock return in get preview function
    httpClientMock.get = jest
      .fn()
      .mockImplementation(() => Promise.resolve({ ok: true, response: response }));
    const wrapper = mount(
      // put it under Formik to render TriggerExpressions that has Formik fields.
      // rendering TriggerExpressions also require adValues to be passed in
      <CoreContext.Provider value={{ http: httpClientMock }}>
        <Formik>
          <AnomalyDetectorTrigger
            detectorId="tempId"
            adValues={{
              anomalyGradeThresholdValue: 0.7,
              anomalyGradeThresholdEnum: 'ABOVE',
              anomalyConfidenceThresholdValue: 0.7,
              anomalyConfidenceThresholdEnum: 0.7,
            }}
            fieldPath=""
          />
        </Formik>
      </CoreContext.Provider>
    );

    expect(httpClientMock.get).toHaveBeenCalledTimes(1);
    await runAllPromises();

    // without update, we will finish mount before the embedded async AnomalyDetectorData finish mounting
    wrapper.update();

    expect(wrapper.update().find('[data-test-subj~="empty-prompt"]').exists()).toBe(false);
    expect(
      wrapper
        .find('[data-test-subj~="anomalyDetector.anomalyGradeThresholdEnum_conditionEnumField"]')
        .exists()
    ).toBe(true);
    expect(
      wrapper
        .find(
          '[data-test-subj~="anomalyDetector.anomalyConfidenceThresholdValue_conditionValueField"]'
        )
        .exists()
    ).toBe(true);
  });
  test('renders no enabled feature', async () => {
    const response = {
      anomalyResult: {
        anomalies: [],
        featureData: {},
      },
      detector: {
        featureAttributes: [
          {
            featureId: 'TV6fFYYB7j86MXY_Bzh2',
            featureName: 'time',
            featureEnabled: false,
            aggregationQuery: {
              time: {
                max: {
                  field: 'time',
                },
              },
            },
          },
        ],
      },
      error: '',
    };

    // Mock return in get preview function
    httpClientMock.get = jest
      .fn()
      .mockImplementation(() => Promise.resolve({ ok: true, response: response }));
    const wrapper = mount(
      <CoreContext.Provider value={{ http: httpClientMock }}>
        <AnomalyDetectorTrigger detectorId="tempId" />
      </CoreContext.Provider>
    );

    await runAllPromises();

    // without update, we will finish mount before the embedded async AnomalyDetectorData finish mounting
    expect(wrapper.update().find('[data-test-subj~="empty-prompt"]').exists()).toBe(true);
    expect(wrapper.find('.euiButton__text').text()).toEqual('Enable Feature');
    expect(wrapper.update().find('[data-test-subj~="_conditionEnumField"]').exists()).toBe(false);
    expect(wrapper.update().find('[data-test-subj~="_conditionValueField"]').exists()).toBe(false);
  });
  test('renders error', async () => {
    const response = {
      anomalyResult: {
        anomalies: [],
        featureData: {},
      },
      detector: {
        featureAttributes: [
          {
            featureId: 'TV6fFYYB7j86MXY_Bzh2',
            featureName: 'time',
            featureEnabled: true,
            aggregationQuery: {
              time: {
                max: {
                  field: 'time',
                },
              },
            },
          },
        ],
      },
      error: 'request error',
    };

    // Mock return in get preview function
    httpClientMock.get = jest
      .fn()
      .mockImplementation(() => Promise.resolve({ ok: true, response: response }));
    const wrapper = mount(
      // put it under Formik to render TriggerExpressions that has Formik fields.
      // rendering TriggerExpressions also require adValues to be passed in
      <CoreContext.Provider value={{ http: httpClientMock }}>
        <Formik>
          <AnomalyDetectorTrigger
            detectorId="tempId"
            adValues={{
              anomalyGradeThresholdValue: 0.7,
              anomalyGradeThresholdEnum: 'ABOVE',
              anomalyConfidenceThresholdValue: 0.7,
              anomalyConfidenceThresholdEnum: 0.7,
            }}
            fieldPath=""
          />
        </Formik>
      </CoreContext.Provider>
    );

    expect(httpClientMock.get).toHaveBeenCalledTimes(1);
    await runAllPromises();

    // without update, we will finish mount before the embedded async AnomalyDetectorData finish mounting
    wrapper.update();

    expect(wrapper.update().find('[data-test-subj~="empty-prompt"]').exists()).toBe(false);
    expect(
      wrapper
        .find('[data-test-subj~="anomalyDetector.anomalyGradeThresholdEnum_conditionEnumField"]')
        .exists()
    ).toBe(true);
    expect(
      wrapper
        .find(
          '[data-test-subj~="anomalyDetector.anomalyConfidenceThresholdValue_conditionValueField"]'
        )
        .exists()
    ).toBe(true);
  });
  test('feature has priority over preview error', async () => {
    const response = {
      anomalyResult: {
        anomalies: [],
        featureData: {},
      },
      detector: {
        featureAttributes: [
          {
            featureId: 'TV6fFYYB7j86MXY_Bzh2',
            featureName: 'time',
            featureEnabled: false,
            aggregationQuery: {
              time: {
                max: {
                  field: 'time',
                },
              },
            },
          },
        ],
      },
      error: 'request error',
    };

    // Mock return in get preview function
    httpClientMock.get = jest
      .fn()
      .mockImplementation(() => Promise.resolve({ ok: true, response: response }));
    const wrapper = mount(
      // put it under Formik to render TriggerExpressions that has Formik fields.
      // rendering TriggerExpressions also require adValues to be passed in
      <CoreContext.Provider value={{ http: httpClientMock }}>
        <Formik>
          <AnomalyDetectorTrigger
            detectorId="tempId"
            adValues={{
              anomalyGradeThresholdValue: 0.7,
              anomalyGradeThresholdEnum: 'ABOVE',
              anomalyConfidenceThresholdValue: 0.7,
              anomalyConfidenceThresholdEnum: 0.7,
            }}
            fieldPath=""
          />
        </Formik>
      </CoreContext.Provider>
    );

    expect(httpClientMock.get).toHaveBeenCalledTimes(1);
    await runAllPromises();

    // without update, we will finish mount before the embedded async AnomalyDetectorData finish mounting
    wrapper.update();

    // without update, we will finish mount before the embedded async AnomalyDetectorData finish mounting
    expect(wrapper.update().find('[data-test-subj~="empty-prompt"]').exists()).toBe(true);
    expect(wrapper.find('.euiButton__text').text()).toEqual('Enable Feature');
    expect(wrapper.update().find('[data-test-subj~="_conditionEnumField"]').exists()).toBe(false);
    expect(wrapper.update().find('[data-test-subj~="_conditionValueField"]').exists()).toBe(false);
  });
});
