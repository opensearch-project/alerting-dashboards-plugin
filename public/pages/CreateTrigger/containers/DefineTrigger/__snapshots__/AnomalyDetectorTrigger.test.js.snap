// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnomalyDetectorTrigger renders no detector id 1`] = `
<div
  style="padding:0px 10px"
>
  <div
    style="border-radius:5px;padding:10px;border:1px solid #D9D9D9;height:250px;width:100%;display:flex;align-items:center;justify-content:center"
  >
    <div
      class="euiEmptyPrompt"
      style="max-width:45em"
    >
      <span
        class="euiTextColor euiTextColor--subdued"
      >
        <div
          class="euiText euiText--medium"
        >
          <div
            class="euiText euiText--medium"
          >
            You must specify a detector.
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`AnomalyDetectorTrigger renders no feature 1`] = `
<div
  style="padding:0px 10px"
>
  <div
    style="border-radius:5px;padding:10px;border:1px solid #D9D9D9;height:250px;width:100%"
  >
    <div
      class="euiEmptyPrompt"
      data-test-subj="empty-prompt"
      style="max-width:45em"
    >
      <span
        class="euiTextColor euiTextColor--subdued"
      >
        <div
          class="euiText euiText--medium"
        >
          <div
            class="euiText euiText--medium"
          >
            No features have been added to this anomaly detector. A feature is a metric that is used for anomaly detection. A detector can discover anomalies across one or more features.
          </div>
        </div>
      </span>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionColumn euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <a
            class="euiButton euiButton--primary"
            data-test-subj="createButton"
            href="anomaly-detection-dashboards#/detectors/tempId/features"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Add Feature
              </span>
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
